### PR TITLE
Move write cache to from datasaver to dataset

### PR DESCRIPTION
**Pull request by jenshnielsen**
First step in creating a more clean interface between the dataset and the datasaver

### CODE REVIEW

File: qcodes/dataset/data_set.py [hunk 1]
```diff
@@ -1167,6 +1176,194 @@ def __repr__(self) -> str:
 
         return "\n".join(out)
 
+    def _enqueue_results(
```
Comment (jenshnielsen) [inline qcodes/dataset/data_set.py:1179]: Perhaps we should make this a public api?

File: qcodes/dataset/measurements.py [hunk 2]
```diff
@@ -669,8 +478,7 @@ def __enter__(self) -> DataSaver:
         self.datasaver = DataSaver(
                             dataset=self.ds,
                             write_period=self.write_period,
-                            interdeps=self._interdependencies,
-                            write_in_background=self._write_in_background)
```
Comment (jenshnielsen) [inline qcodes/dataset/measurements.py:673]: This is tecnially an API break but I don't think anyone is inializing the datasaver directly

Comment (astafan8) [reply qcodes/dataset/measurements.py:673]: exactly. as per our docs, inititlzng DataSaver is NOT a pulbic API :)

File: qcodes/dataset/data_set.py (hunk 1 above)
Comment (astafan8) [reply qcodes/dataset/data_set.py:1179]: that's probably fair, espcially if it's guarded by checking what state the dataset is in - enqueue should only work in the "running" state similar to add_results.

On the other hand, the interfaces of "add_results" and "enqueue results" differ significantly - add_results does not know about qcodes parameters and the fact that things need to unpacked, which enqueue_results knows. So, i would refrain from making it public just for now.

Comment (astafan8) [reply qcodes/dataset/data_set.py:1179]: on the other hand, there's also a publci `add_result_to_queue` :) 

UPDATE: had a look, and that method is NOT used by any code :) moreover, it's behavior is totally included in the behavior of `add_result`. So, either in this or another PR, we need to remove this `add_result_to_queue` method.

Comment (astafan8) [reply qcodes/dataset/data_set.py:1179]: my conclusion - let's keep `_enqueue_results` private for now, and regardless of that guard it with checks for the "running" state of the DataSet.

File: qcodes/dataset/data_set.py [hunk 3]
```diff
@@ -1167,6 +1176,194 @@ def __repr__(self) -> str:
...
+                if value.shape:
+                    res_list += [{param.name: number} for number in value]
+                else:
+                    res_list += [{param.name: complex(value)}]
+            else:
+                res_list += [{param.name: value}]
+
+        return res_list
+
+    def flush_data_to_database(self, block: bool = False) -> None:
```
Comment (astafan8) [inline qcodes/dataset/data_set.py:1336]: shall we keep this private? my thinking is if it's private now, then if we decide to change/remove/move it, we don't need to do deprecation cycles and stuff.

**Pull request merged**